### ISSUE TITLE
Question / Comment: How to test code?

### ISSUE THREAD

**Issue by divchowbey** [pymupdf/PyMuPDF#594]
Hello! I am using PyMuPDF to extract fillable PDFs with text and checkbox inputs. I am having great success with the software, and my next step is to validate the code's accuracy. I am wondering if there is a better way of testing than picking some PDFs and manually confirming the outputs with expected inputs on the form.

From your experiences of working on this library, would you have any recommendations or best practices on how to test the extraction code?

Kind regards

**Comment by JorjMcKie** [comment]
> if there is a better way of testing than picking some PDFs ...

Not really. I have occasionally seen tools which do a similar thing (cannot recall names at the moment), so you may consider comparing their output with yours.
Then you also might create a script which as a first step **_fills_** a PDF based on prepared input input values (e.g. coded in JSON, key-value pairs representing fieldname and value), then run your extractor and compare its output.

It always is a good idea to have different _documented_ testing strategies for unit, string, technical, functional and acceptance test phases (not all of these may be present or required in every situation). Or whether this is a regression test against a predefined test bed after changes.
Depending on this, the effort / testing volume varies a lot. E.g. in **unit test** you probably just check whether all the different field types are handled correctly at all and that this also works for multipage documents and fieldname hierarchies. In technical test, major considerations could be whether document with different PDF versions and formats (e.g. linear - fast web access formats) are handled, too. Etc.
Excuse my verbosity (in professional life I've also had quality assurance functions 😉), but defining a test approach covering these things and **agreeing** it with your users is absolutely critical to limit testing efforts. Otherwise you never know how many hundreds of PDFs you need for confirmation ...
